File: OCamlDream/lib/service.ml
open Models
open Lwt.Infix

module Service = struct
  let handle_db_error err =
    let msg = Caqti_error.show err in
    Lwt.return (Error msg)

  let create_user pool req =
    Caqti_lwt.Pool.use (fun (module C : Caqti_lwt.CONNECTION) ->
      Repository.UserRepo.insert_user (module C) req.username req.email
    ) pool >>= function
    | Ok (Ok user) -> Lwt.return (Ok user)
    | Ok (Error e) -> handle_db_error e
    | Error e -> handle_db_error e

  let list_users pool =
    Caqti_lwt.Pool.use (fun (module C : Caqti_lwt.CONNECTION) ->
      Repository.UserRepo.list_users (module C)
    ) pool >>= function
    | Ok (Ok users) -> Lwt.return (Ok users)
    | Ok (Error e) -> handle_db_error e
    | Error e -> handle_db_error e

  let get_user pool id =
    Caqti_lwt.Pool.use (fun (module C : Caqti_lwt.CONNECTION) ->
      Repository.UserRepo.find_user (module C) id
    ) pool >>= function
    | Ok (Ok user) -> Lwt.return (Ok user)
    | Ok (Error e) -> handle_db_error e
    | Error e -> handle_db_error e

  let update_user pool id req =
    match req.username with
    | None -> get_user pool id (* No update needed *)
    | Some username ->
        Caqti_lwt.Pool.use (fun (module C : Caqti_lwt.CONNECTION) ->
          Repository.UserRepo.update_user (module C) id username
        ) pool >>= function
        | Ok (Ok user) -> Lwt.return (Ok user)
        | Ok (Error e) -> handle_db_error e
        | Error e -> handle_db_error e

  let delete_user pool id =
    Caqti_lwt.Pool.use (fun (module C : Caqti_lwt.CONNECTION) ->
      Repository.UserRepo.delete_user (module C) id
    ) pool >>= function
    | Ok (Ok ()) -> Lwt.return (Ok ())
    | Ok (Error e) -> handle_db_error e
    | Error e -> handle_db_error e

  let create_task pool user_id req =
    Caqti_lwt.Pool.use (fun (module C : Caqti_lwt.CONNECTION) ->
       Repository.UserRepo.find_user (module C) user_id >>= function
       | Error e -> Lwt.return (Error e)
       | Ok None -> Lwt.return (Ok None)
       | Ok (Some _) ->
           Repository.TaskRepo.insert_task (module C) user_id req.title req.description >>= function
           | Error e -> Lwt.return (Error e)
           | Ok task -> Lwt.return (Ok (Some task))
    ) pool >>= function
    | Ok (Ok (Some task)) -> Lwt.return (Ok (Some task))
    | Ok (Ok None) -> Lwt.return (Ok None) (* User not found *)
    | Ok (Error e) -> handle_db_error e
    | Error e -> handle_db_error e

  let list_tasks pool user_id =
    Caqti_lwt.Pool.use (fun (module C : Caqti_lwt.CONNECTION) ->
       Repository.UserRepo.find_user (module C) user_id >>= function
       | Error e -> Lwt.return (Error e)
       | Ok None -> Lwt.return (Ok None)
       | Ok (Some _) ->
           Repository.TaskRepo.list_tasks (module C) user_id >>= function
           | Error e -> Lwt.return (Error e)
           | Ok tasks -> Lwt.return (Ok (Some tasks))
    ) pool >>= function
    | Ok (Ok (Some tasks)) -> Lwt.return (Ok (Some tasks))
    | Ok (Ok None) -> Lwt.return (Ok None) (* User not found *)
    | Ok (Error e) -> handle_db_error e
    | Error e -> handle_db_error e

  let get_task pool id =
    Caqti_lwt.Pool.use (fun (module C : Caqti_lwt.CONNECTION) ->
      Repository.TaskRepo.find_task (module C) id
    ) pool >>= function
    | Ok (Ok task) -> Lwt.return (Ok task)
    | Ok (Error e) -> handle_db_error e
    | Error e -> handle_db_error e

  let update_task pool id req =
    Caqti_lwt.Pool.use (fun (module C : Caqti_lwt.CONNECTION) ->
      Repository.TaskRepo.find_task (module C) id >>= function
      | Error e -> Lwt.return (Error e)
      | Ok None -> Lwt.return (Ok None)
      | Ok (Some current_task) ->
          let title = Option.value req.title ~default:current_task.title in
          let description = Option.value req.description ~default:current_task.description in
          Repository.TaskRepo.update_task (module C) id title description
    ) pool >>= function
    | Ok (Ok task) -> Lwt.return (Ok task)
    | Ok (Error e) -> handle_db_error e
    | Error e -> handle_db_error e

  let patch_task_done pool id =
    Caqti_lwt.Pool.use (fun (module C : Caqti_lwt.CONNECTION) ->
      Repository.TaskRepo.mark_done (module C) id
    ) pool >>= function
    | Ok (Ok task) -> Lwt.return (Ok task)
    | Ok (Error e) -> handle_db_error e
    | Error e -> handle_db_error e

  let delete_task pool id =
     Caqti_lwt.Pool.use (fun (module C : Caqti_lwt.CONNECTION) ->
      Repository.TaskRepo.delete_task (module C) id
    ) pool >>= function
    | Ok (Ok ()) -> Lwt.return (Ok ())
    | Ok (Error e) -> handle_db_error e
    | Error e -> handle_db_error e

end

File: Bun/src/service.ts
import { UserRepository, TaskRepository } from "./repository";
import { User, Task } from "./types";

export class UserService {
    private repo: UserRepository;

    constructor(repo: UserRepository) {
        this.repo = repo;
    }

    createUser(name: string, email: string): User {
        return this.repo.create({ name, email });
    }

    getAllUsers(): User[] {
        return this.repo.getAll();
    }

    getUserById(id: number): User | null {
        return this.repo.getById(id);
    }

    updateUser(id: number, name: string, email: string): User | null {
        return this.repo.update(id, { name, email });
    }

    deleteUser(id: number): boolean {
        return this.repo.delete(id);
    }
}

export class TaskService {
    private taskRepo: TaskRepository;
    private userRepo: UserRepository;

    constructor(taskRepo: TaskRepository, userRepo: UserRepository) {
        this.taskRepo = taskRepo;
        this.userRepo = userRepo;
    }

    createTask(userId: number, title: string, description: string): Task {
        const user = this.userRepo.getById(userId);
        if (!user) {
            throw new Error("User not found");
        }
        return this.taskRepo.create({ user_id: userId, title, description, is_done: false });
    }

    getTasksByUserId(userId: number): Task[] {
        return this.taskRepo.getByUserId(userId);
    }

    getTaskById(id: number): Task | null {
        return this.taskRepo.getById(id);
    }

    updateTask(id: number, title: string, description: string, is_done: boolean): Task | null {
        // Need to fetch existing to know user_id if not provided, but update assumes replacement of fields?
        // Let's assume partial update is not what's asked, but the interface takes all fields.
        // However, repo update takes task object. We might need to fetch first to preserve user_id if we don't pass it.
        // But repository update query doesn't touch user_id.
        // So we can just pass the fields to update.
        // Wait, repository update needs a Task object, which has user_id.
        // But the SQL in repository `UPDATE tasks SET title = $title...` does NOT update user_id.
        // So we can pass 0 for user_id in the repo call safely if we trust the SQL.
        return this.taskRepo.update(id, { user_id: 0, title, description, is_done });
    }

    markTaskDone(id: number): Task | null {
        return this.taskRepo.markDone(id);
    }

    deleteTask(id: number): boolean {
        return this.taskRepo.delete(id);
    }
}
